include/dgl/sampler.h

```diff
@@ -56,13 +56,16 @@ class SamplerOp {
    * \param num_hops the number of hops to sample neighbors.
    * \param expand_factor the max number of neighbors to sample.
    * \param add_self_loop whether to add self loop to the sampled subgraph
+   * \param probability the transition probability (float/double).
    * \return a NodeFlow graph.
    */
-  static NodeFlow NeighborUniformSample(const ImmutableGraph *graph,
-                                        const std::vector<dgl_id_t>& seeds,
-                                        const std::string &edge_type,
-                                        int num_hops, int expand_factor,
-                                        const bool add_self_loop);
+  template<typename ValueType>
+  static NodeFlow NeighborSample(const ImmutableGraph *graph,
+                                 const std::vector<dgl_id_t>& seeds,
+                                 const std::string &edge_type,
+                                 int num_hops, int expand_factor,
+                                 const bool add_self_loop,
+                                 const ValueType *probability);
```

Comment:
does it matter to accept double floating-points for probability?

Reply:
I think it's OK to support both if the code does not become significantly complicated and the performance does not hurt.

Reply:
template in general increases the binary code and compliation time. Also, a large portion of code has to be in the header file. I would suggest to not use template if necessary.

MXNet is an example of overusing templates. I'm hoping that DGL doesn't do the same again.

Reply:
If we were to support multiple data types with the same logic (for instance, int32/int64 graphs, float32/float64 probabilities), then templates seem to be one of the better solutions we have for now.

Other choices include
* [Using `#define`'s all over the place](https://stackoverflow.com/questions/10950828/simulation-of-templates-in-c-for-a-queue-data-type)
* Write your own template (e.g. in Jinja) and render them with a helper script into C code with different data types ([`numpy`](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/scalarmath.c.src)'s solution).

Reply:
Template is definitely better than macro. If you want to support multiple data types, template is definitely a way to go. The question is: do we really need to support both single-precision and double-precision for transition matrix? To support both, we have to put a large amount of code in template. I'm wondering how much supporting both affects the sampling precision?

Reply:
There shouldn't be a lot of code in template.  @jermainewang mentioned that the arrays should be in NDArrays instead of raw pointers, and I agree with that.  With the dispatchers in `aten` I think the only place where we would have templates is the `ArrayHeap`.  I'm leaving the refactor to another PR.

Reply:
@zheng-da, we could solve the binary size problem by using declaration in header file and explicit full specialization in other .cc units. This practice makes it not necessary to inline everything for template functions. For example, suppose you have following:

```c++
template <typename T>
void Foo() {
  // very large function body that you don't want to inline
}

template <typename T>
void Bar() {
  Foo<T>();  // Unfortunately Foo is being used in many other function such as Bar here.
}
```
We could separate the declaration and definition by:
```c++
// foo.h
template <typename T> void Foo();  // only declaration
```
```c++
// foo.cc
template <typename T>
void Foo() {
  // very large function body
}

// explicit specialization for the type that is used
template Foo<int32_t>();
template Foo<int64_t>();
...
```
```c++
// bar.cc
#include "foo.h"
template <typename T>
void Bar() {
  Foo<T>();  // use Foo here will not inline the function.
}
```
This also makes compilation faster and you could even split the template definitions in multiple translation units to parallelize it (the trick used in BinaryOp kernels).

Reply:
The templated code includes `SampleSubgraph`, `GetNonUniformSample` and `ArrayHeap`, so it's still significant.

I'm not trying to block the PR. I'm just saying that we should use template with caution and don't repeat what MXNet did. If you think supporting both sing precision and double precision are definitely necessary, I'm fine with the current implementation.

@jermainewang I know template specialization helps us to move implementation to .cc file and split the template definitions in multiple files. Are you sure it accelerates compilation? I'm not familiar with how compilers deal with template specialization.

Reply:
Let push through this. I think we will gradually move sampling algorithm into `aten` namespace in the future. You could view sampling as a type of sparse operators. As a result, there will be some refactoring going on so if it is currently not too much an issue, let's merge this PR so the functionality is there. What do you think?

Reply:
sounds good to me.